### DISCUSSION TITLE
Include Process Console Output

### DISCUSSION THREAD

**Discussion by BlythMeister** [cake-build/cake#3458] [Ideas]
When exit codes show process errored, it would be great to have the actual console output included in the exception.

Currently, there is no easy way to know in the calling code what caused the non zero exit code.  You know something gone wrong, but no idea what. - granted, its printed in the build log.  But you are not able to take any additional action based on the type of failure (e.g. conditional retry, ignoring the error or doing something different as a fallback)

For example.

Calling the DockerBuild alias, it uses the standard run in Tool.cs.
If this exits with a non zero code, you get an exception thrown, but there is no way to know what happened to cause the error.
DockerBuild is just 1 example...it could be any other alias of built in that uses Tool.cs to start a process.

**Comment by pascalberger** [comment]
ToolSettings (the base class for all built-in and addin tool aliases), currently doesn't support output redirection. This is tracked with https://github.com/cake-build/cake/issues/2460 and https://github.com/cake-build/cake/issues/2461.

Workaround for the moment is to use `StartProcess` alias instead of the specialized tool aliases to run external tools, which allows to redirect standard and error output.

**Reply by BlythMeister** [reply]
I think I understand so rather than me using Cake.Docker 
Use StartProcess to call docker myself?

I don't think the issues you link are what I am after really.  I was more thinking along the lines that currently, when there is a non-zero exit code you get an exception thrown.
That exception could be enriched with the console output.
So in theory, a small change on that base class which anything can leverage.

**Reply by pascalberger** [reply]
> I think I understand so rather than me using Cake.Docker
> Use StartProcess to call docker myself?

Exactly

> I don't think the issues you link are what I am after really. I was more thinking along the lines that currently, when there is a non-zero exit code you get an exception thrown.
> That exception could be enriched with the console output.

IMHO this is exactly what https://github.com/cake-build/cake/issues/2460 is about. Or did I miss something?

**Reply by BlythMeister** [reply]
Sorry, I stand corrected #2460 is what I was thinking of.
